Give the exif data back to pictures

bpater-tp opened the pull request:
They were not written (iOS) to file or destroyed (Android) with FixOrientationAndResizeAsync. I have reverted back to ResizeAsync as I think the library should not do that. It should be app duty to rotate picture where needed, and library should be transparent.

Fixes #184 #185 

Changes Proposed in this pull request:
- store exif data on both iOS and Android, and not let it destroyed when resizing
- extract exif on iOS when photo is selected from gallery
- do not rotate pictures anymore - give to app what it receives from camera
- allow write the gps location to exif, as on iOS and some Android phones it is not available from camera

jamesmontemagno commented on src/Media.Plugin.Android/MediaImplementation.cs:
```diff
@@ -86,7 +86,8 @@ public MediaImplementation()
                 try
                 {
-                    await FixOrientationAndResizeAsync(media.Path, options.PhotoSize, options.CompressionQuality, options.CustomPhotoSize);
+                    //await FixOrientationAndResizeAsync(media.Path, options.PhotoSize, options.CompressionQuality, options.CustomPhotoSize);
```
Is there a reason you are no longer fixing orientation? This is the most requested feature to auto rotate the image. This will only resize it.

jamesmontemagno commented on src/Media.Plugin.Android/MediaImplementation.cs:
```diff
@@ -170,9 +171,13 @@ public MediaImplementation()
             try
             {
-                await FixOrientationAndResizeAsync(media.Path, options.PhotoSize, options.CompressionQuality, options.CustomPhotoSize);
+                var exif = new ExifInterface(media.Path);
```
Same here it is now only resizing? Not fixing orientation?

bpater-tp replied:
I've made rotation default with option to turn it off on demand.

jamesmontemagno commented on src/Media.Plugin.Abstractions/MediaStoreOptions.cs:
```diff
@@ -208,6 +208,10 @@ public int CompressionQuality
+        /// <summary>
+        /// Store provided location
+        /// </summary>
+        public Location UseLocation { get; set; }
```
I don't think this is the correct name. Shouldn't it just be "Location".

UseLocation makes it sound like a boolean

bpater-tp replied:
changed

jamesmontemagno commented on src/Media.Plugin.iOS/PhotoLibraryAccess.cs:
```diff
@@ -0,0 +1,58 @@
+                });
+                while (!completed)
+                {
+                    Thread.Sleep(10);
```
This seems dangerous....

jamesmontemagno replied:
There is no need for this since it is async....

bpater-tp replied:
The case is I have no better idea of how to handle this. Problem here is with `RequestContentEditingInput` method that returns immediately but the actual data are returned lated with the inline callback. If I let Task.Run() finish the data are gone. If you have some better idea how to do it please let me know.

jamesmontemagno commented on src/Media.Plugin.iOS/PhotoLibraryAccess.cs:
```diff
@@ -0,0 +1,58 @@
+            var image = PHAsset.FetchAssets(urls, new PHFetchOptions()).firstObject as PHAsset;
+            await Task.Run(() =>
+            {
+                Task.Run(() =>
```
Why are their 2 of them in here (Task.Run that is)

bpater-tp replied:
removed one

jamesmontemagno commented:
I see your comments on it the rotation. Let's make that an option though. The default for rotation can be true, and you can override.

bpater-tp commented:
Hi. Thanks for comments, I'll look into them early next week. It has been a long day already.

jamesmontemagno commented:
Lovely, i will take a look at these changes soon.

bpater-tp commented:
I have just pushed things today :)

jamesmontemagno commented on src/Media.Plugin.Android/MediaImplementation.cs:
```diff
@@ -604,45 +622,57 @@ public Task<bool> ResizeAsync(string filePath, PhotoSize photoSize, int quality,
         }
 
-        static int GetRotation(string filePath)
+        private string coordsToDMS(double coord)
```
This should be ```CoordinatesTo... whatever DMS is...``` should be more descritive of a title, not sure what it is doing.

bpater-tp replied:
DegreeMinutesSeconds ;) But in fact it is converting coordinates stored as `double` to exif datatype called `RATIONAL`. I've changed the name accordingly.

jamesmontemagno commented on src/Media.Plugin.iOS/MediaPickerDelegate.cs:
```diff
@@ -329,7 +352,18 @@ private bool GetShouldRotate6(UIDeviceOrientation orientation)
+                var success = SaveImageWithMetadata(image, quality, meta, path);
+                if (!success)
+                {
+                    image.AsJPEG(quality).Save(path, true);
```
Should we still save it without metadata?

bpater-tp replied:
I think so. Otherwise it will return nothing...

bpater-tp replied:
I mean - call will return path but there will be no file.

jamesmontemagno commented on src/Media.Plugin.iOS/PhotoLibraryAccess.cs:
```diff
@@ -0,0 +1,55 @@
+            };
+            NSUrl[] urls = { url };
+            var image = PHAsset.FetchAssets(urls, new PHFetchOptions()).firstObject as PHAsset;
+            await Task.Run(() =>
```
This method scares me a lot, seems like it could just sit here and spin forever. Is this documented somewhere you can share with me.

jamesmontemagno replied:
I think there is no time out either so that bothers me too

bpater-tp replied:
ok, this is now rewritten, so there's no sleep or task.delay. It uses task completion mechanism and timeout.

bpater-tp replied:
okay. after digging for a while I have found even better way to fetch the images on iOS. just pushed it. it's now synchronous and does not require that tasks awaiting/canceling/timeouting magic. I hope you like it better as I do. It is smaller and much cleaner.

bpater-tp replied:
after review is finished but before (I still hope for) merging I'd rebase so there are no unnecessary commits with rejected ideas.

jamesmontemagno commented on src/Media.Plugin.iOS/PhotoLibraryAccess.cs:
```diff
@@ -0,0 +1,55 @@
+                });
+                while (!completed)
+                {
+                    Thread.Sleep(10);
```
Thread sleeps scare me too. Task.Delay is a bit better IMHO too.

bpater-tp replied:
see above comment

503geek commented:
Will this allow a choice of what EXIF data is included/removed? I am using the plug-in in a production app and my client specifically requires that EXIF location is not included in the captured photos.

bpater-tp commented:
@503geek no, it's for adding missing gps location. sorry, not this time ;) This kind of feature can be added, but I would prefer to finish one think at a time.

bpater-tp commented:
@jamesmontemagno  are there any other changes that are required?

jamesmontemagno commented:
I need to look into it next week and do testing and resolve the conflicts, but probably should be fine.

jamesmontemagno commented:
@bpater-tp can you check my changes I merged in to make sure I didn't mess anything up and then I will pull it in.

bpater-tp commented:
@jamesmontemagno the changes look ok.

jamesmontemagno commented:
I may have merged incorrect.... i see some issues here: https://ci.appveyor.com/project/JamesMontemagno/mediaplugin/build/2.7.0.32-beta

jamesmontemagno commented:
I changed the brand and going to try to see what happened tomorrow.

bpater-tp commented:
I have rebased everything from my branch with your master (also rebased several commits together to make it easier). Fixed indents, etc. It seems to work now although I have only tested it briefly. Please take a look if it's ok now.

jamesmontemagno commented:
Yay! Thanks!
